src/mds/objects.py

```diff
@@ -391,43 +392,63 @@ async def delete_object(
     guid: str, request: Request, token: HTTPAuthorizationCredentials = Security(bearer)
 ) -> JSONResponse:
     """
-    Delete the metadata for the specified object. Remove the object from
-    existing bucket location(s) by proxying to fence DELETE /data/file_id.
-    Uses the response status code from fence to determine whether user has
+    Delete the metadata for the specified object and also delete the record from indexd. 
+    [Optional] Remove the object from existing bucket location(s) by proxying to
+    fence DELETE /data/file_id by using an additional query parameter `delete_file_locations`.  
+    Uses the response status code from fence/indexd to determine whether user has
     permission to delete metadata.
 
     Args:
         guid (str): indexd GUID or alias, or MDS key
         request (Request): starlette request (which contains reference to FastAPI app)
     Returns:
         204: if record and metadata are deleted
-        403: if fence returns a 403 unauthorized response
-        500: if fence does not return 204 or 403 or there is an error deleting metadata
+        403: if fence/indexd returns a 403 unauthorized response
+        500: if fence/indexd does not return 204 or 403 or there is an error deleting metadata
     """
+    delete_file_locations = False
+    if "delete_file_locations" in request.query_params:
+        if request.query_params["delete_file_locations"]:
+            raise HTTPException(
+            HTTP_400_BAD_REQUEST, f"Query param `delete_file_locations` should not contain any value"
+        )
+        delete_file_locations = True
+    svc_name = "fence" if "delete_file_locations" in request.query_params else "indexd"
     try:
-        fence_endpoint = urljoin(config.DATA_ACCESS_SERVICE_ENDPOINT, f"data/{guid}")
         auth_header = str(request.headers.get("Authorization", ""))
         headers = {"Authorization": auth_header}
-        fence_response = await request.app.async_client.delete(
-            fence_endpoint, headers=headers
-        )
-        fence_response.raise_for_status()
+        if delete_file_locations:
+            fence_endpoint = urljoin(config.DATA_ACCESS_SERVICE_ENDPOINT, f"data/{guid}")
+            response = await request.app.async_client.delete(
+                fence_endpoint, headers=headers
+            )  
+        else:
+            rev = await get_indexd_revision(guid, request)
+            indexd_endpoint = urljoin(config.INDEXING_SERVICE_ENDPOINT, f"index/{guid}")
+            response = await request.app.async_client.delete(
+                indexd_endpoint, params = {"rev": rev}, headers=headers
+            )
+        response.raise_for_status()
         await (
-            Metadata.delete.where(Metadata.guid == guid)
-            .returning(*Metadata)
-            .gino.first()
+        Metadata.delete.where(Metadata.guid == guid)
```

Comment:
the MDS deletion is in the same `try/except` block as the rest, and we're missing the logic we talked about for the SDK (re-create what was deleted if part of the deletion fails)

Reply:
Actually no! I did some analysis again and realized that it was a deliberate choice to not build that. 
It is indeed a good catch that the metadata delete is in the same `try/catch` block as other requests. I can take that out and put it after the `except` block.  
However, I don't think we need to re-create anything here because, if there is an exception during `fence DELETE` or `indexd DELETE`, it means that these `DELETE` options haven't been successful and thereby we never reach the `Metadata DELETE` step but directly to HTTPException response to the user.  Correct me if I'm wrong.